### DISCUSSION TITLE
What is .../_next/data/development/...json doing in getServerSideProps?

### DISCUSSION THREAD

**Discussion by BlakeBrown** [vercel/next.js#15787] [Help] [answered]
I'm having trouble understanding how getServerSideProps() behaves. Let's say I navigate to `/decks`

```js
export async function getServerSideProps(context) {
 const url = `${context.req.protocol}://${context.req['hostname']}/api${context.req.originalUrl}`;
 const raw_response = await fetch(url);
 ...
}
```

If I reload the page (SSR), the url is http://localhost:3000/api/decks ✅ 
If I navigate to the page via `Router.push('/decks')` on the client, the url is http://localhost:3000/api/_next/data/development/decks.json ❌ 

A few questions:
1. What is this json file?
2. How do I get the original URL?
3. How do I share data between pages when doing client-side rendering?

I've read through the documentation and haven't been able to find anything on this, would be really helpful to document `getServerSideProps()` 🙏

**Comment by BlakeBrown** [comment]
Bump. Updated the question to make it more clear 😄

**Comment by jamesmosier** [comment, marked as answer]
1. When you navigate to a page via client side routing, Next executes your `getServerSideProps` function by calling this as an API request. [`getServerSideProps` docs](https://nextjs.org/docs/basic-features/data-fetching#only-runs-on-server-side).
2. What do you mean by "How do I get the original URL?"? You could use environment variables to set the domain of your API routes. Or better yet you could put your logic directly into `getServerSideProps` instead of calling an API route. [This section of the `getStaticProps` docs touches on this topic](https://nextjs.org/docs/basic-features/data-fetching#write-server-side-code-directly).
3. This is something you'd have to handle within your own code. Whether it is localStorage or sessionStorage to store data between pages or via querystring parameters. Next doesn't have any mechanism for passing data between pages like that.

**Reply by jamesmosier** [reply]
From the [docs](https://nextjs.org/docs/basic-features/data-fetching#only-runs-on-server-side):

> When you request this page on client-side page transitions through next/link or next/router, Next.js sends an API request to the server, which runs getServerSideProps. It’ll return JSON that contains the result of running getServerSideProps, and the JSON will be used to render the page. All this work will be handled automatically by Next.js, so you don’t need to do anything extra as long as you have getServerSideProps defined.

So said another way, when you access a page that utilizes `getServerSideProps` via a full page refresh (i.e. going directly to the url in your browser) Next will execute the logic within `getServerSideProps` on the server. But if you navigate to this same page but via `next/link` or `next/router` then the page is not rendered on the server, therefore Next needs a way to still execute your `getServerSideProps` logic. So this is where this JSON API comes into play. Next will automatically create this JSON API for you and it will be called when you navigate to this page client side. The `.json` extension is just to signify that the HTTP resource returns JSON.

For example, this page:

```jsx
export default function Product(props){
  return <div>{props.product.name}</div>
}

export async function getServerSideProps(context) {
  return {
    props: { product: { name: 'Shirt' } }, 
  }
}
```

when executed on the server the `getServerSideProps` will be run on the server and `props.product.name` will have a value when the page is server rendered.

But when navigated via `<Link href="/product">go to my product</Link>`, an API will be called (`product.json`) and return:

```json
{
  "product": { "name": "Shirt" }
}
```

because this page is client rendered and the page needs a way to get the data, so it makes an HTTP request to get the data.

---

As far as how to get the current URL, if you are using dynamic routes then `context.params` would work to know what the route is otherwise why not just hardcode the route (instead of `context.req.originalUrl`)?

**Reply by BlakeBrown** [reply]
Ah so getServerSideProps should always be expected to return static data? I shouldn't be making API requests which return dynamic content because the JSON file can't change?

BTW REALLY APPRECIATE THIS 🙏 you've gone above and beyond any answer I was expecting, thank you so much

**Reply by binoiii** [reply]
@jamesmosier 

> But if you navigate to this same page but via next/link or next/router then the page is not rendered on the server, therefore Next needs > a way to still execute your getServerSideProps logic. So this is where this JSON API comes into play. Next will automatically create this > JSON API for you and it will be called when you navigate to this page client side.

On my machine I encountered 301 on the initial request of the JSON API. Any thoughts?

**Comment by BlakeBrown** [comment]
Appreciate the response @jamesmosier! 

1. What does "Next executes your getServerSideProps function by calling this as an API request" mean? The documentation is lacking explanation as to where the JSON file comes from or how to access it. Is Next creating JSON files on my server?
2. I meant how to get the URL after navigating. I think `window.location.href` will probably work. So we need to duplicate requests in `getServerSideProps()` into `componentDidMount()` to fetch data for the page?
3. Seriously?! Wow that's bad 😅 `<Link>` or `Router` should probably support the ability to pass props to a page. Would be happy to help with this when I'm more familiar with the framework.

**Comment by Fi1osof** [comment]
How can i switch off this API requests on client-side? I want this calls on server-side and do not on client-side. 
Thanks!

**Reply by jamesmosier** [reply]
You cannot switch it off. The reason being is that when you navigate using `next/link` then the page wouldn't have the data that is returned in `getServerSideProps`. `getServerSideProps` is primarily meant to send data from the server to the client.

**Reply by Fi1osof** [reply]
@jamesmosier , thanks!
I think should use getInitialProps instead.

**Comment by josteph** [comment]
So how does this connect with ISR pages? 

Considering if we put the fetch at `getStaticProps` and then we navigate using for example `router.push`, the data that contained in `.json` will simply not be updated too by ISR?

The question is more like, does the `revalidate` strategy actually update the `.json` file too?

**Comment by sevencai** [comment]
> All this work will be handled automatically by Next.js, so you don’t need to do anything extra as long as you have getServerSideProps defined.

From this [doc](https://nextjs.org/docs/basic-features/data-fetching#only-runs-on-server-side) mentioned by jamesmosier , It says we don't need to do anything extra when using `getServerSideProps`.  But If we're trying to get cookies, I found that we need to specify where to get the cookie from, otherwise we can't access the right cookies. It takes me some time to find the problem and I am wondering if anyone else has  encountered similar problems. I am  quite confused about the explaination in the doc.

```javascript
export function getToken(req) {
  // get cookies from document
  if (isClientSide()) {
    return xxxx
  }
  
  // get cookies from req object
  return req.cookies[xxx]
}
```
